File: 46. Permutations.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) <= 1:
            return [nums]
        ans = [[]]
        for i in range(len(nums)):
            le = len(ans)
            l = 0
            while l < le:
                l1 = ans.pop(0)
                l1.insert(0, nums[i])
                ans.append(l1.copy())
                for j in range(len(l1) - 1):
                    l1[j], l1[j+1] = l1[j+1], l1[j]
                    ans.append(l1.copy())
                l += 1
        return ans


sol = Solution()
print(sol.permute([]))
print(sol.permute([11]))
print(sol.permute([11, 22]))
print(sol.permute([11, 22, 33]))
print(sol.permute([11, 22, 33, 44]))

File: 46. Permutations.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int> &nums) {
        vector<bool> flag(nums.size(), false);
        vector<vector<int>> res;
        vector<int> curr;
        backtracking(nums, curr, flag, res);
        return res;
    }

    void backtracking(const vector<int> &nums, vector<int> &curr, vector<bool> &flag, vector<vector<int>> &res) {
        if (curr.size() == nums.size()) {
            res.push_back(curr);
            return;
        }
        for (int i=0; i < nums.size(); ++i) {
            if (flag[i]) continue;
            curr.push_back(nums[i]);
            flag[i] = true;
            backtracking(nums, curr, flag, res);
            curr.pop_back();
            flag[i] = false;
        }
    }
};

int main() {
    Solution a;
    vector<int> v{1,2,3};
    vector<vector<int>> ans = a.permute(v);
    for (auto &vec: ans) {
        for (auto element: vec) {
            cout << element << ' ';
        }
        cout << endl;
    }
    return 0;
}
